components/style/atomic_refcell.rs

```diff
@@ -2,121 +2,358 @@
+                // space. Before panicking, decrement the refcount to leave things
+                // in a consistent immutable-borrow state.
+                //
+                // This will never happen in a real program.
+                borrow.fetch_sub(1, atomic::Ordering::AcqRel);
+                panic!("too many immutable borrows");
+            } else if new >= MAX_FAILED_BORROWS {
+                // During the mutable borrow, an absurd number of threads have
+                // incremented the refcount and panicked. To avoid hypothetically
+                // wrapping the refcount, we abort the process once a certain
+                // threshold is reached.
+                //
+                // This will never happen in a real program.
+                //
+                // FIXME(bholley): We can't currently whole-process abort in
+                // stable Rust, but it's coming soon. We should fix this when
+                // https://github.com/rust-lang/rust/issues/37838 hits release.
+                panic!("Too many failed borrows");
+            } else {
+                // This is the normal case, and the only one which should happen
+                // in a real program.
+                panic!("already mutably borrowed");
+            }
+        }
+
+        AtomicBorrowRef { borrow: borrow }
     }
 }
 
-impl<'a, T: 'a + Debug> Debug for AtomicRefMut<'a, T> {
-    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
-        write!(f, "{:?}", self.0.deref())
+impl<'b> Drop for AtomicBorrowRef<'b> {
+    #[inline]
+    fn drop(&mut self) {
+        let old = self.borrow.fetch_sub(1, atomic::Ordering::AcqRel);
+        // This assertion is technically incorrect in the case where another
+        // thread hits the hypothetical overflow case, since we might observe
+        // the refcount before it fixes it up (and panics). But that never will
+        // never happen in a real program, and this is a debug_assert! anyway.
+        debug_assert!(old & HIGH_BIT == 0);
     }
 }
 
-impl<T> AtomicRefCell<T> {
-    pub fn new(value: T) -> Self {
-        AtomicRefCell(RwLock::new(value))
+struct AtomicBorrowRefMut<'b> {
+    borrow: &'b AtomicUsize,
+}
+
+impl<'b> Drop for AtomicBorrowRefMut<'b> {
+    #[inline]
+    fn drop(&mut self) {
+        let old = self.borrow.swap(0, atomic::Ordering::AcqRel);
+        // The old value may be slightly higher than HIGH_BIT if another thread
+        // tried to immutably borrow and panicked.
+        debug_assert!(old >= HIGH_BIT);
     }
-    pub fn borrow(&self) -> AtomicRef<T> {
-        AtomicRef(self.0.try_read().expect("already mutably borrowed"))
+}
+
+impl<'b> AtomicBorrowRefMut<'b> {
+    #[inline]
+    fn new(borrow: &'b AtomicUsize) -> AtomicBorrowRefMut<'b> {
+        // Use compare-and-swap to avoid corrupting the immutable borrow count
+        // on illegal mutable borrows.
+        let old = borrow.compare_and_swap(0, HIGH_BIT, atomic::Ordering::AcqRel);
```

Comment:
Additionally, you should use the `compare_exchange_strong` method, which allows you to specify a different memory ordering on failure (`Relaxed` is enough in this case).